homa_incoming.c

```diff
@@ -683,20 +683,20 @@ void homa_resend_pkt(struct sk_buff *skb, struct homa_rpc *rpc,
 
 	if (!homa_is_client(rpc->id) && rpc->state != RPC_OUTGOING) {
 		/* We are the server for this RPC. */
-		if ((rpc->msgin.length - rpc->msgin.bytes_remaining) >= rpc->msgin.granted) {
-			/* We've received everything that we've granted, so either the server 
-			   is busy with an influx of messages exceeding max_incoming, or a 
-			   grant packet has been dropped. In either case, the server should 
-			   initiate RESEND
-			   */
+		if ((rpc->msgin.length - rpc->msgin.bytes_remaining) 
+				>= rpc->msgin.granted) {
+			/* We've received everything that we've granted, so likely we
+			 * are busy with an influx of messages exceeding max_incoming.
+			 * We should send a BUSY right now and initiate a RESEND
+			 * later when we are free.
+			 */
 			tt_record2("sending BUSY from resend, id %d, state %d",
 					rpc->id, rpc->state);
 			homa_xmit_control(BUSY, &busy, sizeof(busy), rpc);
-		}
-		else{
-			/* We haven't received everything we've granted so somehow DATA 
-			   packets have been lost. Issue RESEND immediately
-		   	*/
+		} else {
+			/* We haven't received everything we've granted; so somehow 
+			 * DATA packets have been lost. Issue RESEND immediately.
+		   	 */
 			homa_get_resend_range(&rpc->msgin, h);
 			h->priority = rpc->hsk->homa->num_priorities -1;
 			homa_xmit_control(RESEND, h, sizeof(h), rpc);
```

Comment:
I think this code will work, but I'd suggest a different approach. homa_get_resend already checks rpc->msgin.granted, so I'd suggest letting it do that work, rather than repeating it here. I think you can eliminate the test on line 686 and just invoke homa_get_resend immediately. If we've received all of the bytes that have been granted for the message, h->resend_length will be zero after the call. So, just check that after homa_get_resend returns and send BUSY instead of RESEND if it's zero. This approach is better because it encapsulates all knowledge of whether a RESEND is needed in homa_get_resend.